### PR TITLE
feat(eslint-plugin): add rule `use-unknown-in-catch-callback-variables`

### LINKED ISSUE #7526: Rule proposal: always use type `(error: unknown) => void` with promise.catch
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

TypeScript has the [`useUnknownInCatchVariables`](https://www.typescriptlang.org/tsconfig#useUnknownInCatchVariables) option to use `unknown` with catch clauses instead of `any`. However, `promise.catch` is stuck with using an `any` type no matter how you configure this option.

I'd like a rule to enforce using the type `unknown` with the caught error handler for promises. This could either use type inference or just lint every method named `.catch`.

### Fail Cases

```typescript
promise.catch((error) => { ... });
promise.catch((error: Error) => { ... });
promise.catch((error: any) => { ... });
```

### Pass Cases

```typescript
promise.catch((error: unknown) => { ... });
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by kirkwaiblinger**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7526 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This change implements the rule proposal requiring variables in `.catch()` callbacks to have type `unknown`. The rule fundamentally works by flagging on the condition
- the first argument to a `.catch()` call (on a Thenable) is a function
- And the first parameter to said function is both _present_, and not definitely the intrinsic `unknown` type (according to the type checker, not syntax analysis).

Note that the backing issue mentions the return type, but any return type is legitimate in a `.catch` callback, so this rule does not check the return type.

Fixes for cases that are analyzable syntactically have been implemented as editor suggestions.

### Limitations
- Because this is flagging based on the types of the handler, not the syntax, it probably cannot be used to force you to provide `err: unknown` if you _have_ modified the gobal type declarations for `.catch()`.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:110]: Question for maintainers - I more or less copied this pattern for dealing with rest args from no-misused-promsies:
https://github.com/typescript-eslint/typescript-eslint/blob/289ee8885ce26714296d8258abb12a2fa2c38d61/packages/eslint-plugin/src/rules/no-misused-promises.ts#L544-L589

Is there/should there be a shared utility to "go into rest arg"?

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:363]: I'm pretty sure this function is not sound, but I also haven't figured out a way to break it. 

Is there a shared utility or better way to do this?

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:346]: This is copied from prefer-find, a la https://github.com/typescript-eslint/typescript-eslint/issues/8323. Should we fix that for this PR?

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:109]: Question for maintainers - I more or less copied this pattern for dealing with rest args from no-misused-promsies:
https://github.com/typescript-eslint/typescript-eslint/blob/289ee8885ce26714296d8258abb12a2fa2c38d61/packages/eslint-plugin/src/rules/no-misused-promises.ts#L544-L589

Is there/should there be a shared utility to "go into rest arg"?

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:362]: I'm pretty sure this function is not sound, but I also haven't figured out a way to break it. 

Is there a shared utility or better way to do this?

File: packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md
Comment (kirkwaiblinger) [inline packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md:40]: Add example with present but incorrect type annotation on handler (`.catch((err: Error) => {}))`)

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:109]: Yeah I think that's reasonable, +1! Let's try it out 😄 

[Praise] +1 for suggesting!

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:176]: [Docs] Nit, I might remove some of these <20 line comments that just restate the code. Up to you.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
This looks great, awesome job on all the edge cases & whatnot!

Left a few comments inline, requesting changes. But really solid all-up - no major notes! 🙌

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:356]: This was a super helpful tip. The function is now a one liner :) And has no assertions/throws/instanbul-ignores.

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:36]: touche. Forgot that that was not breaking

File: packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md
Comment (kirkwaiblinger) [reply packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md:59]: makes sense. Made changes around this.

Comment (kirkwaiblinger) [reply packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md:59]: Made some changes in this area

File: packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md
Comment (kirkwaiblinger) [reply packages/eslint-plugin/docs/rules/use-unknown-in-catch-callback-variable.md]: Yeah that definitely makes sense. Made significant changes in this area.

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: Changed this up a bit. Still a bit hairy so you may want to validate the change

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: Changed this up.

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: Totally. Changed all of these!

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: "Normal" here is meant to be "normal" argument rather than spread argument, as opposed to implying that this the "normal" case for the rule to check (although tbh that's true too haha). But that is certainly ambiguous from the wording!

Reworded to make it explicit that I'm talking about non-spread arguments

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:1]: I would like to say that the syntax in that first one is very upsetting to me 😆 

Added several test cases in these areas.

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: I went the route of separating object and array destructuring into two separate types and being explicit about what invalid assumptions are in each. Feel free to review and approve or disapprove :)

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: changed

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:108]: Hmm, unfortunately there's not that much duplication that I could see how to obviously eliminate... But I made a small dent 🤷

File: packages/eslint-plugin/src/configs/all.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/configs/all.ts:1]: [Question] Err, what's going on here? Prettier shenanigans?

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts:7]: [Naming] Everywhere else lowercases this:

```suggestion
import * as tsutils from 'ts-api-utils';
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
This looks great to me, nicely done! 👏 

Since this rule touches a lot of logical areas, I think I'd want another review from @typescript-eslint/triage-team before merging. Adding `1 approval` in case someone has time in the next few weeks.

File: packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts]: changed

File: packages/eslint-plugin/src/configs/all.ts
Comment (kirkwaiblinger) [reply packages/eslint-plugin/src/configs/all.ts:1]: Weirdly, on my branch, configs are not generating with the correctly resolved prettier config. Not sure why. 

I do note, though, that there has been some activity in the config generation files lately, so I may have just branched from a commit where there was an oddity. In any case, fixed the formatting, and it doesn't appear to a problem on latest on "main" either.

**Review by JoshuaKGoldberg — DISMISSED** [review]
...I clicked the wrong button last time, sorry. 😄 

Screenshot of my review requesting changes

File: packages/eslint-plugin/src/configs/strict-type-checked.ts
Comment (bradzacher) [inline packages/eslint-plugin/src/configs/strict-type-checked.ts:79]: just double checking/confirming we want to add this to the strict config before the next major?
i don't think we've ever expressly discussed our strategy for the strict configs.
they're not technically included in our [versioning policy](https://typescript-eslint.io/users/versioning#eslint-plugin) for a major release if they're changed - but I wonder if adding a brand new rule in a minor might be a bit disruptive for users?

cc @JoshuaKGoldberg

**Review by bradzacher — COMMENTED** [review]
quick eyeball - haven't got time for a full review at this second.

File: packages/eslint-plugin/src/configs/strict-type-checked.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/configs/strict-type-checked.ts:79]: Yup confirmed, we've done this before (#7957, #8011). Since the strict config is explicitly out of semver, users are opting into this.

**Comment by fregante** [comment]
In some of my projects, I already set the type of the exception to `unknown` via global.d.ts types. Can this rule detect and accept that? This is basically a lib issue

**Comment by kirkwaiblinger** [comment]
@fregante 

> In some of my projects, I already set the type of the exception to `unknown` via global.d.ts types. Can this rule detect and accept that? This is basically a lib issue

To be honest, I didn't test this, since the purpose of this rule existing (per the issue) is for the people who don't modify the lib, and I don't really remember since I worked on this a while back.  I did put a note in the docs that that's not the intended use case, though.   https://main--typescript-eslint.netlify.app/rules/use-unknown-in-catch-callback-variable#when-not-to-use-it

Feel free to share what you find if you test it!

**Pull request merged**
